### ISSUE TITLE
Inference failure since slice patterns does not serve as type expections

### ISSUE THREAD

**Issue by oxalica** [rust-lang/rust-analyzer#16609]
Types of `a` and `b` in this code below cannot be inferred by rust-analyzer, but it compiles fine with rustc.
```rust
// edition = "2021"
fn main() {
    let [a, b] = vec![1u32, 2u32].try_into().unwrap();
    //   ^ {unknown}
}
```

**rust-analyzer version**: 2024-02-19

**rustc version**: rustc 1.76.0 (07dca489a 2024-02-04)

**relevant settings**: N/A

**Comment by Veykril** [comment]
Hmm, `vec![1u32, 2u32].try_into()` has type `Result<[{unknown}], <Vec<u32, Global> as TryInto<[{unknown}]>>::Error>` here, adding the type ascription `let [a, b]: [u32; 2] = vec![1u32, 2u32].try_into().unwrap();` makes us resolve the `try_into` result correctly though

**Comment by oxalica** [comment]
> adding the type ascription

Yes. But `rustc` can infer that without explicit type ascription.

**Comment by Veykril** [comment]
I am aware, I was just trying to reduce the root problem. Oh thanks for that link, would make sense that we are missing that if its such a recent inference feature.
